private/buf/buflsp/server.go

```diff
@@ -215,9 +215,22 @@ func (s *server) Formatting(
 		return nil, err
 	}
 
+	nodeInfo := file.fileNode.NodeInfo(file.fileNode)
 	return []protocol.TextEdit{
 		{
-			Range:   infoToRange(file.fileNode.NodeInfo(file.fileNode)),
+			Range: protocol.Range{
+				// The formatter always returns the entire file; make sure we
+				// represent the edit as the entire range (which may not match up to
+				// the file's NodeInfo.Start()).
+				Start: protocol.Position{
+					Line:      0,
+					Character: 0,
+				},
+				End: protocol.Position{
+					Line:      uint32(nodeInfo.End().Line) - 1,
+					Character: uint32(nodeInfo.End().Col) - 1,
+				},
+			},
```

Comment:
This is still potentially incorrect, due to trailing comments.

Instead, given that parsing is a relatively intense operation, I would prefer that we simply count the number of
lines in file.fileNode and the number of characters on the last line, and construct a Range using that. Don't bother with the spans in the AST, since they clearly don't reflect what we actually mean.

Also, include the following comment:
```go
// XXX: The current compiler does not expose a span for the full file. Instead of
// potentially undershooting the correct span (which can cause comments at the
// start and end of the file to be duplicated), we instead manually count up the
// number of lines in the file. This is comparatively cheap, compared to sending the
// entire file over a domain socket.
```

Also, please file a bug against protocompile to include a way to retrieve a span of the whole file.

Reply:
FWIW, it doesn't seem like the issue extends to trailing comments, maybe because [`NodeInfo.End()` has special handling for EOF](https://pkg.go.dev/github.com/bufbuild/protocompile/ast#NodeInfo.End)?

> I would prefer that we simply count the number of lines in file.fileNode and the number of characters on the last line, and construct a Range using that.

I was looking around at the FileNode APIs and didn't find much exposed that would help with that; mind pointing me to how that would be done / make that change yourself? (Happy for you to take this over; my `protocompile` knowledge is weak 😅.)

Reply:
I'm juggling a bunch of PRs already and am about to go on vacation, so I prefer you do it if only to make sure it gets done. Also, good excuse to grow your knowledge!

Anyways, the right way to do this:
```go
var lastLine, lastLineStart int
for i := 0; i < len(file.text); i++ {
  // NOTE: we are iterating over bytes, not runes.
  if file.text[i] == '\n' {
    lastLine++
    lastLineStart = i + 1 // Skip the \n.
  }
}
lastChar := len(file.text[lastLineStart:]) - 1 // Bytes, not runes!
```

Reply:
https://github.com/bufbuild/buf/pull/3424/commits/0057f0db8ca9b51227164c886a6338c50ea2fece, PTAL!

Reply:
> Also, please file a bug against protocompile to include a way to retrieve a span of the whole file.

https://github.com/bufbuild/protocompile/issues/360!